File: packages/sdk/src/resources/memory-spaces.ts
/**
 * Memory Spaces Resource
 *
 * Resource-based SDK interface for shared memory space management.
 * Memory spaces allow agents to collaborate within shared knowledge containers.
 */

import { ResponseShapeError } from "../errors.js";
import { unwrapData, unwrapDataObject, requireArray, type JsonObject } from "../validate.js";
import { BaseResource } from "./base.js";

const RESOURCE = "memory-spaces";

/**
 * Narrow a named object field of an unwrapped memory-spaces envelope (e.g.
 * `{ data: { space } }`), throwing a structured {@link ResponseShapeError}
 * when the field is missing or not an object.
 */
function requireMember<T>(data: JsonObject, key: string, route: string): T {
  const value = data[key];
  if (!value || typeof value !== "object") {
    throw new ResponseShapeError(
      `Unexpected ${route} response shape (expected { data: { ${key} } })`,
      route,
      RESOURCE,
    );
  }
  return value as T;
}

// ============================================================================
// Types
// ============================================================================

export type MemorySpacePermission = "read" | "write" | "admin";

export interface MemorySpace {
  id: string;
  title: string;
  description: string | null;
  visibility: "private" | "shared";
  nodeType: "memory_space";
  createdAt: string;
  updatedAt: string;
}

export interface MemorySpaceWithMembers extends MemorySpace {
  members: MemorySpaceMember[];
}

export interface MemorySpaceMember {
  agentId: string;
  permission: MemorySpacePermission;
  joinedAt: string;
}

export interface CreateMemorySpaceParams {
  title: string;
  description?: string;
  visibility?: "private" | "shared";
  initialMembers?: Array<{ agentId: string; permission: MemorySpacePermission }>;
}

export interface ListMemorySpacesParams {
  agentId?: string;
}

export interface JoinMemorySpaceParams {
  agentId: string;
  permission: MemorySpacePermission;
}

// ============================================================================
// API Response Types
// ============================================================================

interface ApiSuccessResponse<T> {
  data: T;
  meta?: {
    pagination?: {
      total: number;
      limit: number;
      offset?: number;
      hasMore: boolean;
    };
  };
}

// ============================================================================
// Memory Spaces Resource
// ============================================================================

/**
 * Memory Spaces Resource - manages shared memory spaces for multi-agent collaboration.
 */
export class MemorySpacesResource extends BaseResource {
  /**
   * List memory spaces, optionally filtered by agent membership.
   */
  async list(params?: ListMemorySpacesParams): Promise<MemorySpace[]> {
    const query = new URLSearchParams();
    if (params?.agentId) {
      query.set("agentId", params.agentId);
    }

    const queryString = query.toString();
    const path = `/v1/memory-spaces${queryString ? `?${queryString}` : ""}`;

    const route = `GET ${path}`;
    const response = await this.request<ApiSuccessResponse<{ spaces: MemorySpace[] }>>(
      "GET",
      path
    );
    return requireArray<MemorySpace>(
      unwrapDataObject(response, route, RESOURCE).spaces,
      route,
      RESOURCE,
    );
  }

  /**
   * Create a new memory space.
   */
  async create(params: CreateMemorySpaceParams): Promise<MemorySpace> {
    const route = "POST /v1/memory-spaces";
    const response = await this.request<ApiSuccessResponse<{ space: MemorySpace }>>(
      "POST",
      "/v1/memory-spaces",
      params
    );
    return requireMember<MemorySpace>(unwrapDataObject(response, route, RESOURCE), "space", route);
  }

  /**
   * Get a memory space by ID, including its members.
   */
  async get(spaceId: string): Promise<MemorySpaceWithMembers> {
    const route = `GET /v1/memory-spaces/${encodeURIComponent(spaceId)}`;
    const response = await this.request<ApiSuccessResponse<{ space: MemorySpaceWithMembers }>>(
      "GET",
      `/v1/memory-spaces/${encodeURIComponent(spaceId)}`
    );
    return requireMember<MemorySpaceWithMembers>(unwrapDataObject(response, route, RESOURCE), "space", route);
  }

  /**
   * Join a memory space as an agent with a given permission level.
   */
  async join(spaceId: string, params: JoinMemorySpaceParams): Promise<MemorySpaceMember> {
    const route = `POST /v1/memory-spaces/${encodeURIComponent(spaceId)}/join`;
    const response = await this.request<ApiSuccessResponse<{ member: MemorySpaceMember }>>(
      "POST",
      `/v1/memory-spaces/${encodeURIComponent(spaceId)}/join`,
      params
    );
    return requireMember<MemorySpaceMember>(unwrapDataObject(response, route, RESOURCE), "member", route);
  }

  /**
   * Leave a memory space (remove an agent from the space).
   */
  async leave(spaceId: string, agentId: string): Promise<{ removed: true }> {
    const query = new URLSearchParams();
    query.set("agentId", agentId);
    const qs = query.toString();
    const response = await this.request<ApiSuccessResponse<{ removed: true }>>(
      "DELETE",
      `/v1/memory-spaces/${encodeURIComponent(spaceId)}/leave${qs ? `?${qs}` : ""}`
    );
    return unwrapData(response, `DELETE /v1/memory-spaces/${encodeURIComponent(spaceId)}/leave`, RESOURCE);
  }
}

File: packages/sdk-python/engram/resources/memory_spaces.py
"""Memory space resources for the Engram SDK.

Resource-based interface for shared memory space management. Memory spaces
allow agents to collaborate within shared knowledge containers.

Mirrors the TypeScript SDK's ``MemorySpacesResource``
(``packages/sdk/src/resources/memory-spaces.ts``).

These endpoints use the standard ``{ data }`` envelope, but the inner payload
is keyed by member name (``{ data: { space } }``, ``{ data: { spaces } }``,
``{ data: { member } }``). Each method unwraps ``data`` and then the named
member, matching the TS contract exactly.
"""
from __future__ import annotations

from typing import Any, List, Optional, TYPE_CHECKING
from urllib.parse import quote

from engram._base import BaseResource, SyncBaseResource
from engram.types.memory_spaces import (
    CreateMemorySpaceParams,
    JoinMemorySpaceParams,
    MemorySpace,
    MemorySpaceMember,
    MemorySpaceWithMembers,
)

if TYPE_CHECKING:
    from engram.client import AsyncEngramClient, EngramClient


def _unwrap_member(response: Any, key: str) -> Any:
    """Unwrap ``response["data"][key]`` from a memory-spaces envelope.

    The memory-spaces routes wrap their payload as ``{ data: { <key> } }``
    (e.g. ``{ data: { space } }``). This mirrors the TS ``unwrapDataObject``
    followed by the named-member read.
    """
    data = response["data"]
    return data[key]


class MemorySpacesResource(BaseResource):
    """Async resource for shared memory spaces."""

    async def list(self, agent_id: Optional[str] = None) -> List[MemorySpace]:
        """List memory spaces, optionally filtered by agent membership.

        Args:
            agent_id: Filter to spaces the given agent is a member of.

        Returns:
            List of memory spaces.
        """
        qs: dict[str, str] = {}
        if agent_id is not None:
            qs["agentId"] = agent_id
        response = await self._request(
            "GET", "/v1/memory-spaces", params=qs if qs else None
        )
        spaces = _unwrap_member(response, "spaces")
        return [MemorySpace.model_validate(item) for item in spaces]

    async def create(self, params: CreateMemorySpaceParams) -> MemorySpace:
        """Create a new memory space.

        Args:
            params: Title, optional description/visibility, and initial members.

        Returns:
            The created memory space.
        """
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = await self._request("POST", "/v1/memory-spaces", body)
        return MemorySpace.model_validate(_unwrap_member(response, "space"))

    async def get(self, space_id: str) -> MemorySpaceWithMembers:
        """Get a memory space by ID, including its members.

        Args:
            space_id: The memory space ID.

        Returns:
            The memory space with its members.
        """
        response = await self._request(
            "GET", f"/v1/memory-spaces/{quote(space_id, safe='')}"
        )
        return MemorySpaceWithMembers.model_validate(_unwrap_member(response, "space"))

    async def join(
        self, space_id: str, params: JoinMemorySpaceParams
    ) -> MemorySpaceMember:
        """Join a memory space as an agent with a given permission level.

        Args:
            space_id: The memory space ID.
            params: The agent ID and permission level.

        Returns:
            The created membership record.
        """
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = await self._request(
            "POST", f"/v1/memory-spaces/{quote(space_id, safe='')}/join", body
        )
        return MemorySpaceMember.model_validate(_unwrap_member(response, "member"))

    async def leave(self, space_id: str, agent_id: str) -> dict[str, Any]:
        """Leave a memory space (remove an agent from the space).

        Args:
            space_id: The memory space ID.
            agent_id: The agent to remove.

        Returns:
            ``{"removed": True}`` on success.
        """
        response = await self._request(
            "DELETE",
            f"/v1/memory-spaces/{quote(space_id, safe='')}/leave",
            params={"agentId": agent_id},
        )
        return response["data"]


class SyncMemorySpacesResource(SyncBaseResource):
    """Sync resource for shared memory spaces."""

    def list(self, agent_id: Optional[str] = None) -> List[MemorySpace]:
        """List memory spaces, optionally filtered by agent membership.

        Args:
            agent_id: Filter to spaces the given agent is a member of.

        Returns:
            List of memory spaces.
        """
        qs: dict[str, str] = {}
        if agent_id is not None:
            qs["agentId"] = agent_id
        response = self._request(
            "GET", "/v1/memory-spaces", params=qs if qs else None
        )
        spaces = _unwrap_member(response, "spaces")
        return [MemorySpace.model_validate(item) for item in spaces]

    def create(self, params: CreateMemorySpaceParams) -> MemorySpace:
        """Create a new memory space.

        Args:
            params: Title, optional description/visibility, and initial members.

        Returns:
            The created memory space.
        """
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = self._request("POST", "/v1/memory-spaces", body)
        return MemorySpace.model_validate(_unwrap_member(response, "space"))

    def get(self, space_id: str) -> MemorySpaceWithMembers:
        """Get a memory space by ID, including its members.

        Args:
            space_id: The memory space ID.

        Returns:
            The memory space with its members.
        """
        response = self._request(
            "GET", f"/v1/memory-spaces/{quote(space_id, safe='')}"
        )
        return MemorySpaceWithMembers.model_validate(_unwrap_member(response, "space"))

    def join(self, space_id: str, params: JoinMemorySpaceParams) -> MemorySpaceMember:
        """Join a memory space as an agent with a given permission level.

        Args:
            space_id: The memory space ID.
            params: The agent ID and permission level.

        Returns:
            The created membership record.
        """
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = self._request(
            "POST", f"/v1/memory-spaces/{quote(space_id, safe='')}/join", body
        )
        return MemorySpaceMember.model_validate(_unwrap_member(response, "member"))

    def leave(self, space_id: str, agent_id: str) -> dict[str, Any]:
        """Leave a memory space (remove an agent from the space).

        Args:
            space_id: The memory space ID.
            agent_id: The agent to remove.

        Returns:
            ``{"removed": True}`` on success.
        """
        response = self._request(
            "DELETE",
            f"/v1/memory-spaces/{quote(space_id, safe='')}/leave",
            params={"agentId": agent_id},
        )
        return response["data"]
